sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReduceNumShufflePartitions.scala

```diff
@@ -76,12 +82,7 @@ case class ReduceNumShufflePartitions(conf: SQLConf) extends Rule[SparkPlan] {
       // `ShuffleQueryStageExec` gives null mapOutputStatistics when the input RDD has 0 partitions,
       // we should skip it when calculating the `partitionStartIndices`.
       val validMetrics = shuffleMetrics.filter(_ != null)
-      // We may get different pre-shuffle partition number if user calls repartition manually.
-      // We don't reduce shuffle partition number in that case.
-      val distinctNumPreShufflePartitions =
-        validMetrics.map(stats => stats.bytesByPartitionId.length).distinct
-
-      if (validMetrics.nonEmpty && distinctNumPreShufflePartitions.length == 1) {
+      if (validMetrics.nonEmpty) {
```

Comment:
@carsonwang is it safe to remove `distinctNumPreShufflePartitions.length == 1` from here? I think @hvanhovell's comment (https://github.com/apache/spark/pull/24978/files#r299396944) still applies here about `Union`. I run into an issue with my plan:
```
Union
:- Project [id_key#236, true AS row_type#249, link#232]
:  +- Filter (isnotnull(min(id_key) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#246) AND (id_key#236 = min(id_key) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#246))
:     +- Window [min(id_key#236) windowspecdefinition(specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS min(id_key) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#246]
:        +- ShuffleQueryStage 5
:           +- Exchange SinglePartition, true
:              +- *(7) Project [id_key#236, link#232]
:                 +- *(7) SortMergeJoin [link#237], [link#232], Inner, (id_key#236 > id_key#230)
:                    :- *(5) Sort [link#237 ASC NULLS FIRST], false, 0
:                    :  +- CoalescedShuffleReader [0]
:                    :     +- ShuffleQueryStage 0
:                    :        +- Exchange hashpartitioning(link#237, 5), true
:                    :           +- *(1) Project [col1#224 AS id_key#236, col2#225 AS link#237]
:                    :              +- *(1) LocalTableScan [col1#224, col2#225]
:                    +- *(6) Sort [link#232 ASC NULLS FIRST], false, 0
:                       +- CoalescedShuffleReader [0]
:                          +- ShuffleQueryStage 1
:                             +- Exchange hashpartitioning(link#232, 5), true
:                                +- *(2) Project [id_key#230, link#232]
:                                   +- *(2) Filter (isnotnull(link#232) AND isnotnull(id_key#230))
:                                      +- *(2) Scan RecursiveReference iter[id_key#230,row_type#231,link#232]
+- Project [id_key#240, new AS new#256, link#241]
   +- SortMergeJoin [id_key#238], [id_key#240], Inner
      :- Sort [id_key#238 ASC NULLS FIRST], false, 0
      :  +- ShuffleQueryStage 4
      :     +- Exchange hashpartitioning(id_key#238, 5), true
      :        +- *(4) Project [id_key#238]
      :           +- *(4) Filter (isnotnull(min(id_key) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#247) AND (id_key#238 = min(id_key) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#247))
      :              +- Window [min(id_key#238) windowspecdefinition(specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS min(id_key) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#247]
      :                 +- CoalescedShuffleReader [0]
      :                    +- ShuffleQueryStage 2
      :                       +- Exchange SinglePartition, true
      :                          +- LocalTableScan <empty>, [id_key#238]
      +- Sort [id_key#240 ASC NULLS FIRST], false, 0
         +- ShuffleQueryStage 3
            +- Exchange hashpartitioning(id_key#240, 5), true
               +- *(3) Project [col1#228 AS id_key#240, col2#229 AS link#241]
                  +- *(3) LocalTableScan [col1#228, col2#229]
```
where `ShuffleQueryStage 5` conflicts with `ShuffleQueryStage 4` and `ShuffleQueryStage 3`.

Reply:
ah `SinglePartition` is an exception. So it's still possible to hit `distinctNumPreShufflePartitions.length > 1` here. Let's add back this check @carsonwang @maryannxue 

thanks for reporting it!

Reply:
I've opened a small PR here: https://github.com/apache/spark/pull/25479 as a follow-up, please let me know if this requires a new ticket.